hoomd/hpmc/IntegratorHPMCMonoGPU.h

```diff
@@ -460,29 +1014,137 @@ void IntegratorHPMCMonoGPU< Shape >::initializeExcellMem()
 
     // reallocate memory
     m_excell_idx.resize(m_excell_list_indexer.getNumElements());
     m_excell_size.resize(num_cells);
+
+    if (this->m_exec_conf->allConcurrentManagedAccess())
+        {
+        // set memory hints
+        auto gpu_map = this->m_exec_conf->getGPUIds();
+        for (unsigned int idev = 0; idev < this->m_exec_conf->getNumActiveGPUs(); ++idev)
+            {
+            cudaMemAdvise(m_excell_idx.get(), sizeof(unsigned int)*m_excell_idx.getNumElements(), cudaMemAdviseSetAccessedBy, gpu_map[idev]);
+            cudaMemAdvise(m_excell_size.get(), sizeof(unsigned int)*m_excell_size.getNumElements(), cudaMemAdviseSetAccessedBy, gpu_map[idev]);
+            CHECK_CUDA_ERROR();
+            }
+        }
+    }
+
+template< class Shape >
+void IntegratorHPMCMonoGPU< Shape >::slotNumTypesChange()
+    {
+    unsigned int old_ntypes = this->m_params.size();
+
+    // skip the reallocation if the number of types does not change
+    // this keeps shape parameters when restoring a snapshot
+    // it will result in invalid coefficients if the snapshot has a different type id -> name mapping
+    if (this->m_pdata->getNTypes() != old_ntypes)
+        {
+        unsigned int ntypes = this->m_pdata->getNTypes();
+
+        // resize array
+        GlobalArray<Scalar> lambda(ntypes*ntypes, this->m_exec_conf);
+        m_lambda.swap(lambda);
+        TAG_ALLOCATION(m_lambda);
+
+        // ntypes counters per GPU, separated by at least a memory page
+        unsigned int pitch = (getpagesize() + sizeof(hpmc_implicit_counters_t)-1)/sizeof(hpmc_implicit_counters_t);
+        GlobalArray<hpmc_implicit_counters_t>(std::max(pitch, this->m_implicit_count.getNumElements()),
+            this->m_exec_conf->getNumActiveGPUs(), this->m_exec_conf).swap(m_implicit_counters);
+        TAG_ALLOCATION(m_implicit_counters);
+
+        if (this->m_exec_conf->allConcurrentManagedAccess())
+            {
+            // memory hint for overlap matrix
+            cudaMemAdvise(this->m_overlaps.get(), sizeof(unsigned int)*this->m_overlaps.getNumElements(), cudaMemAdviseSetReadMostly, 0);
+            CHECK_CUDA_ERROR();
+            }
+        }
+
+    // call base class method
+    IntegratorHPMCMono<Shape>::slotNumTypesChange();
     }
 
 template< class Shape >
 void IntegratorHPMCMonoGPU< Shape >::updateCellWidth()
     {
+    // call base class method
     IntegratorHPMCMono<Shape>::updateCellWidth();
 
-    // changing the cell width means that the particle shapes have changed, assume this invalidates the
-    // image list and aabb tree
-    this->m_image_list_valid = false;
-    this->m_aabb_tree_invalid = true;
-
+    // update the cell list
     this->m_cl->setNominalWidth(this->m_nominal_width);
+
+    // set memory hints
+    cudaMemAdvise(this->m_params.data(), this->m_params.size()*sizeof(typename Shape::param_type), cudaMemAdviseSetReadMostly, 0);
+    CHECK_CUDA_ERROR();
+
+    #if 0
```

Comment:
Is there a need to keep this unused code?

Reply:
hmm, would be good to change `attach_to_stream` to `set_memory_hint` so that appropriate `cudaMemAdviseSetReadMostly` flags can be set on the parameters and its substructures

Reply:
done